File: login/javascript/user_info.js
/*
 *	@Project  : Helios
 *	@Architecture : Kim Bom
 *	signup.js
 *
 *	@Created by KimBom On 2016. 1. 2...
 *	@Copyright (C) 2015 KimBom. All rights reserved.
 */

/**
 * @name CheckBlank
 * @check1 : 공백검사
 * @check2 : ID,PW,EMAIL,CPW,FILE 유효성 검사
 */
function CheckSubmit() {
    var id = document.getElementById("ID").value;
	var pw = document.getElementById("PW").value;
	var name = document.getElementById("NAME").value;
	var guild = document.getElementById("GUILD").value;
	var phone = document.getElementById("PHONE").value;
	var email = document.getElementById("EMAIL").value;
	var file = document.getElementById("FILE").value;
	// 공백검사
    
	if (pw == "" || name == "" || guild == "" || phone == ""
			|| email == "") {
		alert("Please enter all information.");
		return false;
	}
	// 비밀번호 검사
	if (pw_keyup() == false) {
		alert("password error");
		return false;
	}
	// 이메일 검사
	if (em_keyup() == false) {
		alert("email error");
		return false;
	}
	// 확인 비밀번호 검사
	if (cpw_keyup() == false) {
		alert("confirm password error");
		return false;
	}
	// 파일 확장자 검사
    if(file!=""){
        if (file_chk() == false) {
            alert("file upload error");
            return false;
        }
    }
    
	// SHA3 암호화
	var salt=CryptoJS.lib.WordArray.random(32);
	var enc_pw=CryptoJS.SHA3(salt+pw);
	//var enc_id=CryptoJS.SHA3(id);
	document.getElementById("HPW").value = enc_pw;
	document.getElementById("HID").value = id;
	document.getElementById("HSALT").value = salt;
    var default_salt=document.getElementById("DEFAULT_SALT").value;
    var curr_pw=document.getElementById("CURRPW").value;
    var default_pw=CryptoJS.SHA3(default_salt+curr_pw);
    document.getElementById("CURRPW").value=default_pw;
    
	return true;
	
}
/**
 * @name : pw_keyup
 * @check1 : pw length<4
 */
function pw_keyup() {
	var pw = document.getElementById("PW").value;
	var i = document.getElementById("PW");
	if (i.value.length > 0) {
		i.value = i.value.replace(/[^(\d|a-z|A-Z|_)]+/g, '');
	}
	if(pw!=i.value){
		var ivc = document.getElementById("ivchar");
		ivc.style.visibility = "visible";
		return false;
	}else{
		var ivc = document.getElementById("ivchar");
		ivc.style.visibility = "hidden";
	}
	// 길이가 4이하면 에러
	if (pw.length >= 4) {
		var pw4 = document.getElementById("pass4");
		pw4.style.visibility = "hidden";
		return true;
	} else {
		var pw4 = document.getElementById("pass4");
		pw4.style.visibility = "visible";
		return false;
	}
}
/**
 * @name : cpw_keyup
 * @check1 : 비밀번호와 같은지 검사
 */
function cpw_keyup() {
	var pw = document.getElementById("PW").value;
	var cpw = document.getElementById("CPW").value;
	if (cpw == pw) {
		var c_pw = document.getElementById("cpass");
		c_pw.style.visibility = "hidden";
		return true;
	} else {
		var c_pw = document.getElementById("cpass");
		c_pw.style.visibility = "visible";
		return false;
	}
}
/**
 * @name : not empty keyup
 * @check1 : 공백검사
 * @param1 : 속성 ID
 * @param2 : 경고 메세지 ID
 */
function nempty_keyup(param1, param2) {
	var value = document.getElementById(param1).value;
	if (value == "") {
		var bar = document.getElementById(param2);
		bar.style.visibility = "visible";
	} else {
		var bar = document.getElementById(param2);
		bar.style.visibility = "hidden";
	}
}
/**
 * @name : email_keyup
 * @check1 : email check
 */
function em_keyup() {
	var email = document.getElementById("EMAIL").value;
	var regex = /^([\w-]+(?:\.[\w-]+)*)@((?:[\w-]+\.)*\w[\w-]{0,66})\.([a-z]{2,6}(?:\.[a-z]{2})?)$/;
	if (regex.test(email) === false) {
		var em = document.getElementById("emchk");
		em.style.visibility = "visible";
		return false;
	} else {
		var em = document.getElementById("emchk");
		em.style.visibility = "hidden";
		return true;
	}
}
/**
 * @name : tel_keyup
 * @check1 : 숫자만 입력가능(다른문자가 들어오면 제거)
 */
function tel_keyup() {
	var i = document.getElementById("PHONE");
	if (i.value.length > 0) {
		i.value = i.value.replace(/[^\d]+/g, '');
	}
	if (i.value == "") {
		var em = document.getElementById("phonechk");
		em.style.visibility = "visible";
		return true;
	} else {
		var em = document.getElementById("phonechk");
		em.style.visibility = "hidden";
	}
	return true;
}
/**
 * @name : file check
 * @check1 : 확장자검사
 */
function file_chk() {
	var file = document.getElementById("FILE").value;
	var execs = new Array(".jpg", ".gif",  ".bmp", ".png",".jpeg");
	var bSubmitCheck = false;
	if (!file) {
		alert("please select your photo");
		return;
	}
	while (file.indexOf("\\") != -1) {
		file = file.substring(file.indexOf("\\") + 1);
	}
	ext = file.substring(file.indexOf(".")).toLowerCase();
	for (var i = 0; i < execs.length; i++) {
		if (execs[i] == ext) {
			bSubmitCheck = true;
			break;
		}
	}
	if (!bSubmitCheck) {
		alert("enable upload file.\n\n" + (execs.join("  "))
				+ "please select again");
		return false;
	}
	return true;
}
/**
 * @name : file size check
 * @check1 : file size < 4mb
 */
function filesizechk(objFile) {
	var nMaxSize = 4 * 1024 * 1024; // 4 MB
	var nFileSize = objFile.files[0].size;
	if (nFileSize > nMaxSize) {
		document.getElementById("FILE").value = "";
		document.getElementById("fakeFile").value = "";
		alert("Please upload a file less than 4MB");
	} else {
		document.getElementById("fakeFile").value = document
				.getElementById("FILE").value;
	}
}
/**
 * @name : go file
 * @comment : 파일을 클릭하게 합니다.
 */
function goFile() {
	document.getElementById("FILE").click();
}

File: login/user_info.pl
#!/usr/bin/perl
#
#	@Project  : Helios
#	@Architecture : Kim Bom
#	user_info.pl
#
#	@Created by KimBom On 2016. 2. 19...
#	@Copyright (C) 2015 KimBom. All rights reserved.
#
use strict;
use warnings;
use CGI;
use DBI;
use File::Copy;
use Digest::SHA3 qw(sha3_512_hex);
use Crypt::Salt;
require 'info.pl';
require 'aes.pl';

#==============================CGI 작업을 준비합니다.==============================
my $q = new CGI;
my $con = DBI->connect( GetDB(), GetID(), GetPW() );

#==============================Form의 값을 받아옵니다.==============================
my $id          = $q->param('HID');
my $pw          = $q->param('HPW');
my $name        = $q->param('NAME');
my $guild       = $q->param('GUILD');
my $phone       = $q->param('PHONE');
my $email       = $q->param('EMAIL');
my $upload_file = $q->param('FILE');
my $salt        = $q->param('HSALT');
#$id='root';$pw='root';$name='root';$guild='root';$phone='1234';$email='root';$salt='abvs';
my $ids = "";    #데이터베이스의 아이디 목록입니다.

#get cookie
my $enc_name=AES_Encrypt("bluecandle_helios_cookie_id");
chop($enc_name);
chop($enc_name);
chop($enc_name);
my $c=$q->cookie($enc_name);	#cookie
if($c){
	$c=AES_Decrypt($c);
}
my $e_id="";
my $e_name="";
my $e_guild="";
my $e_phone="";
my $e_email="";
my $e_photo="";
my $e_salt="";
if($c){
    my $state=$con->prepare("SELECT ui_id,ui_name,ui_guild,ui_phone,ui_email,ui_photo,ui_salt1 FROM userinfo WHERE ui_id=\'$c\'");
    $state->execute();
    while ( my $row = $state->fetchrow_hashref ) {
        $e_id=$row->{ui_id};
        $e_name=$row->{ui_name};
        $e_guild=$row->{ui_guild};
        $e_phone=$row->{ui_phone};
        $e_email=$row->{ui_email};
        $e_photo=$row->{ui_photo};
        $e_salt=$row->{ui_salt1};
    }
}else{
    print $q->redirect("../index.pl");
}
my $script="";
#==============================회원가입을 완료했을때 호출됩니다..==============================
my $DEBUG=$id;
if ($id) {
    #기존 비밀번호의 매칭을 검사.
    my $default_salt="";
    my $default_pw="";
     my $state=$con->prepare("SELECT ui_pw,ui_salt2 FROM userinfo WHERE ui_id=\'$c\'");
    $state->execute();
    while ( my $row = $state->fetchrow_hashref ) {    
        $default_salt=$row->{ui_salt2};
        $default_pw=$row->{ui_pw};
    }
    my $passchk=sha3_512_hex($default_salt.$q->param('CURRPW'));
    $DEBUG=$q->param('CURRPW');
    if($default_pw ne $passchk){
        $script='<script>alert("current password is wrong!!")</script>';
    }else{
    my $photo_path=$e_photo;
    if($upload_file ne ""){
	   $upload_file =~ m<.+\.(\w+)?$>;    #확장자를 추출합니다.
        my $exe=$1;
       $e_photo =~ m/-([^-]*)-/;
       my $num=$1+1;
	   $photo_path = "$id-$num-.$exe";
	   copy( $upload_file, "photo/$photo_path" );	#파일을 저장합니다.
    }
	my $p_salt=salt(32);
	$pw=sha3_512_hex($p_salt.$pw);
    my $query="UPDATE userinfo set ui_pw=\'$pw\' , ui_name=\'$name\' , ui_guild=\'$guild\' , ui_phone=\'$phone\' ,ui_email=\'$email\',ui_photo=\'$photo_path\'
    ,ui_salt1=\'$salt\',ui_salt2=\'$p_salt\' WHERE ui_id=\'$id\'";
    $DEBUG=$query;
	#my $query = "INSERT INTO userinfo values(\'$id\',\'$pw\',\'$name\',\'$guild\',\'$phone\',\'$email\',\'$photo_path\',\'0\',\'$salt\',\'$p_salt\')";
	$con->do($query);
	$con->disconnect();
	print $q->redirect('../index.pl');
    }
}
else {
	my $query = "SELECT ui_id FROM userinfo";
	my $state = $con->prepare($query);
	$state->execute();
	while ( my @row = $state->fetchrow_array ) {
		$ids = $ids . $row[0] . ",";
	}
	chop($ids);
	$state->finish();
	$con->disconnect();
    
}
#=======================================DEBUG=================
my $visible="hidden";
#==============================CGI 부분==============================72
print $q->header( -charset => "UTF-8" );
print <<EOF
<head>
	<title>BlueCandle</title>
	<link rel="stylesheet" type="text/css" href="css/user_info.css" />
	<script src="javascript/sha3.js"></script>
	<script src="javascript/pbkdf2.js"></script>
	<script src="javascript/user_info.js" type="text/javascript"></script>
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    $script
</head>
<body>
<p>$DEBUG</p>
<section class="back">
		<form action="../index.pl" method="post">
			<input type="submit" id="login"  value="go to main"></input>
		</form>
</selection>
<div class="container">	
	<form action="user_info.pl" method="post" ENCTYPE="multipart/form-data">
    
	<img src="photo/$e_photo" width="200px" height="200px" style="border-radius:20px;position:absolute;left:42%">
    <div class="outer"><div class="inner"><div class="centered-top"></div><div class="centered">
    
		<div class="centered_inner_right"></div>
		<div id="msg" class="centered_inner_right2" >
			
			<div class='warn' id="pass4" style="visibility:$visible;top:226px">Password is too short</div>
			<div class='warn' id="ivchar" style="visibility:$visible;top:186px">Invaild character</div>
			<div class='warn' id="cpass" style="visibility:$visible;top:196px">not matching password</div>
			<div class='warn' id="namechk" style="visibility:$visible;top:206px">input your name</div>
			<div class='warn' id="guildchk" style="visibility:$visible;top:216px">input your guild</div>
			<div class='warn' id="phonechk" style="visibility:$visible;top:226px">input your phone</div>
			<div class='warn' id="emchk" style="visibility:$visible;top:236px">This is not a valid e-mail</div>
		</div>
		<div class="centered_inner_left">
			<p class="leftP">Your best id</p>
			<p class="leftP">Safety password</p>
			<p class="leftP">Confirm password</p>
			<p class="leftP">What`s your name?</p>
			<p class="leftP">Where do you belong?</p>
			<p class="leftP">Phone number</p>
			<p class="leftP">Your most used e-mail</p>
			<p class="leftP">Your pretty face</p>
            <p class="leftP2">Current Password</p>
		</div>
		<section class="webdesigntuts-workshop" >
			<input type="hidden" id="IDS"  value="$ids"></input>
			<input type="hidden" id="HPW" name="HPW" value=""></input>
			<input type="hidden" id="HID" name="HID" value=""></input>
			<input type="hidden" id="HSALT" name="HSALT" value=""></input>
            <input type="hidden" id="DEFAULT_SALT" name="DEFAULT_SALT" value="$e_salt"></input>
            
			<input type="text" id="ID" autocomplete="off" name="ID"  onkeyup="id_keyup()" onblur="id_keyup()" maxlength="64" value="$e_id" readonly="readonly"></input>
			<input type="password" id="PW" autocomplete="off" name="PW"  onkeyup="pw_keyup()" onblur="pw_keyup()" maxlength="64"></input>
			<input type="password" id="CPW" autocomplete="off" name="CPW"  onkeyup="cpw_keyup()" onblur="pw_keyup()" maxlength="64"></input>
			<input type="text" id="NAME" autocomplete="off" name="NAME"  maxlength="64" onkeyup="nempty_keyup('NAME','namechk')" onblur="nempty_keyup('NAME','namechk')" value="$e_name"></input>
			<input type="text" id="GUILD" autocomplete="off" name="GUILD"  maxlength="64" onkeyup="nempty_keyup('GUILD','guildchk')" onblur="nempty_keyup('GUILD','guildchk')" value="$e_guild"></input>
			<input type="text" id="PHONE" autocomplete="off" name="PHONE"  onkeyup="tel_keyup()" onblur="tel_keyup()" maxlength="64" value="$e_phone"></input>
			<input type="email" id="EMAIL" autocomplete="off" name="EMAIL"  onkeyup="em_keyup()" onblur="em_keyup()" maxlength="100" value="$e_email"></input>
			<input type="file" style="display:none" id="FILE" name="FILE" onchange="filesizechk(this)" accept="image/*" value="$e_photo"></input>
	
			<input type="text" id="fakeFile" value="" placeholder="***default***" style="width:200px;" readonly="readonly"></input>
			<input type="button" value="UPLOAD" onclick="goFile()" style="width:90px;"></input>
            <input type="password" id="CURRPW" autocomplete="off" name="CURRPW"  onkeyup="pw_keyup()" onblur="pw_keyup()" maxlength="64"></input>
			<input type="submit" value="Confirm information" onclick="return CheckSubmit();"></input>
		</section>
	</div></div></div></div>
	</form>
	
</body>
</html>
EOF
  ;
